pkg/controller/acmechallenges/update.go

```diff
@@ -0,0 +1,132 @@
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	internalchallenges "github.com/cert-manager/cert-manager/internal/controller/challenges"
+	"github.com/cert-manager/cert-manager/internal/controller/feature"
+	cmacme "github.com/cert-manager/cert-manager/pkg/apis/acme/v1"
+	"github.com/cert-manager/cert-manager/pkg/client/clientset/versioned"
+	utilfeature "github.com/cert-manager/cert-manager/pkg/util/feature"
+	"github.com/cert-manager/cert-manager/test/unit/gen"
+)
+
+type updater interface {
+	update(context.Context, *cmacme.Challenge) (*cmacme.Challenge, error)
+	updateStatus(context.Context, *cmacme.Challenge) (*cmacme.Challenge, error)
+}
+
+type objectUpdater struct {
+	updater
+}
+
+func newObjectUpdater(cl versioned.Interface, fieldManager string) *objectUpdater {
+	o := &objectUpdater{
+		updater: &objectUpdaterDefault{cl: cl},
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(feature.ServerSideApply) {
+		o.updater = &objectUpdaterSSA{
+			fieldManager: fieldManager,
+			cl:           cl,
+		}
+	}
+	return o
+}
+
+// updateObject updates the Finalizers if they have changed and updates the Status if it has changed.
+// Finalizers are updated using the Update method while Status is updated using
+// the UpdateStatus method.
+// Both updates will be attempted, even if one fails, except in the case where
+// one of the updates fails with a Not Found error.
+// If the any of the API operations results in a Not Found error, updateObject
+// will exit without error and the remaining operations will be skipped.
+// Only the Finalizers and Status fields may be modified. If there are any
+// modifications to new object, outside of the Finalizers and Status fields,
+// this function will panic.
+func (o *objectUpdater) updateObject(ctx context.Context, old, new *cmacme.Challenge) error {
+	if !apiequality.Semantic.DeepEqual(
+		gen.ChallengeFrom(old, gen.SetChallengeFinalizers(nil), gen.ResetChallengeStatus()),
+		gen.ChallengeFrom(new, gen.SetChallengeFinalizers(nil), gen.ResetChallengeStatus()),
+	) {
+		panic("only the finalizers and status fields may be modified")
+	}
+
+	var updateFunctions []func() (*cmacme.Challenge, error)
+	if !apiequality.Semantic.DeepEqual(old.Status, new.Status) {
+		updateFunctions = append(
+			updateFunctions,
+			func() (*cmacme.Challenge, error) {
+				obj, err := o.updateStatus(ctx, new)
+				return obj, errors.Wrap(err, "when updating the status")
+			},
+		)
+	}
+	if !apiequality.Semantic.DeepEqual(old.Finalizers, new.Finalizers) {
+		updateFunctions = append(
+			updateFunctions,
+			func() (*cmacme.Challenge, error) {
+				obj, err := o.update(ctx, new)
+				return obj, errors.Wrap(err, "when updating the finalizers")
+			},
+		)
+	}
+	var errors []error
+	for _, f := range updateFunctions {
```

Comment:
Thought: so we update status before spec and it would be possible that a status update succeeds, but a spec update fails.
Can we thus end up in a state where the finalizer did not get applied due to spec apply failure, but a 'final' status did get applied and the challenge potentially gets deleted before cleaning up the resources? Should we apply spec first?

Reply:
I kept it the same order as it was. See the original (deleted) comment elsewhere in this patch.
> // call UpdateStatus first as we may have updated the challenge.status.reason field

If the Update failed to remove the Finalizer, then the Challenge can't be deleted and it would be re-reconciled and the cleanup would be re-attempted.

Reply:
What I am thinking about was a flow for a `Challenge` in 'processing' state and without finalizer:

1. A `Challenge` gets synced
2. Finalizer gets added
3. In the same reconcile loop, the self-check and the authorization succeeds so the `Challenge` enters a final state (valid?) and we are returning from the reconcile
4. The deferred function runs, the call to update status succeeds, but the call to update spec fails
5. The `Challenge` is now in a final state, but does not have a finalizer
6. Now there is race between the challenges controller adding the finalizer in another reconcile loop and the `Challenge` getting garbage collected. If the `Challenge` gets garbage collected first, the resources won't be cleaned up

Let me know if you think this is a potential edge case.

> I kept it the same order as it was

Previously if there was an error when applying the finalizer we would have returned early before processing the `Challenge` I think.

Reply:
The Sync function *always* returns immediately after adding the finalizer. It won't do self-check and accept in the same calll.
And if the deferred Update fails, then it will exit with err and the next Sync will again attempt to add the finalizer.

https://github.com/wallrj/cert-manager/blob/1ade01f8191c97e90d301679409825ba003a7002/pkg/controller/acmechallenges/sync.go#L87-L93

Reply:
Ahh, that's my bad, I didn't spot that it returns after adding the finalizer!